src/runtime/pipeline/pipeline_scheduler.cc

```diff
@@ -26,12 +27,97 @@ namespace runtime {
  * \brief Initialize the pipeline.
  * \param modules The list of graph executor modules.
  * \param pipeline_conf The dependency information of each graph executor module.
+ * \return Return a list of backend runtime module.
  */
-size_t PipelineScheduler::PipelineInit(const std::vector<Module>& modules,
-                                       const PipelineConfig& pipeline_config) {
+std::vector<std::shared_ptr<BackendRuntime>> PipelineScheduler::PipelineInit(
+    const std::vector<Module>& modules, ConfigPipelineExecution pipeline_config) {
+  std::vector<std::shared_ptr<BackendRuntime>> runtimes;
   graph_modules_ = modules;
-  int num_output = pipeline_config.GetGlobalOutputNum();
-  return num_output;
+  for (size_t i = 0; i < graph_modules_.size(); i++) {
+    auto runItem = std::make_shared<BackendRuntime>(graph_modules_[i], i);
+    runtimes.push_back(runItem);
+  }
+  // Initialize the outputs array.
+  auto& global_output_map = pipeline_config.GetGlobalConfigOutputBindings();
+  for (size_t i = 0; i < global_output_map.size(); i++) {
+    if (global_output_map.find(i) == global_output_map.end()) {
+      LOG(FATAL) << "Not find global output index " << i;
+    }
+    ModuleOutputPair& output_pair = global_output_map[i];
+    NDArray output = runtimes[output_pair.mod_idx]->CreateFromOutput(output_pair.output_idx);
+    output_array.push_back(output);
+  }
+  return runtimes;
 }
+
+/*!
+ * \brief Exeute in the serialized mode.
+ * \param runtimes A list of backend runtimes module.
+ * \param pipeline_config The dependency information of each graph executor module.
+ */
+void PipelineScheduler::PipelineRunSerial(
+    const std::vector<std::shared_ptr<BackendRuntime>>& runtimes,
+    ConfigPipelineExecution pipeline_config) {
+  for (size_t i = 0; i < runtimes.size(); i++) {
+    // The offset in vector is the runtime execution order, this offset value should
+    // be same with the the value of "runtime_idx" in runtime.
+    if (static_cast<int>(i) != runtimes[i]->GetModuleIndex()) {
+      LOG(FATAL) << "runtime index " << runtimes[i]->GetModuleIndex()
+                 << " is not same as vector offset value " << i;
```

Comment:
Will the pipeline stop properly (e.g., kill running threads, etc) when encountering an error like this?

Reply:
yes for pipeline running, a stop should get call for stop properly, in this case it is just a serialized mode and be single thread, hence no need stop logic get involed.